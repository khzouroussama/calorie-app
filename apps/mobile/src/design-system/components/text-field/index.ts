export {
  TextField,
  TextFieldProps,
  TextFieldAccessoryProps,
} from './text-field';
