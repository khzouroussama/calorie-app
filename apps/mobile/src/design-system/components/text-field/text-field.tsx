import React, {
  ComponentType,
  forwardRef,
  Ref,
  useImperativeHandle,
  useRef,
} from 'react';
import {
  StyleProp,
  TextInput,
  TextInputProps,
  TextStyle,
  TouchableOpacity,
  View,
  ViewStyle,
} from 'react-native';
import { colors, spacing, typography } from '../../theme';
import { Typography, TypographyProps } from '../typography';

export interface TextFieldAccessoryProps {
  style: StyleProp<ViewStyle>;
  status: TextFieldProps['status'];
  multiline: boolean;
  editable: boolean;
}

export interface TextFieldProps extends Omit<TextInputProps, 'ref'> {
  status?: 'error' | 'disabled';
  label?: TypographyProps['text'];
  LabelTextProps?: TypographyProps;
  helper?: TypographyProps['text'];
  HelperTextProps?: TypographyProps;
  placeholder?: TypographyProps['text'];
  style?: StyleProp<TextStyle>;
  containerStyle?: StyleProp<ViewStyle>;
  inputWrapperStyle?: StyleProp<ViewStyle>;
  RightAccessory?: ComponentType<TextFieldAccessoryProps>;
  LeftAccessory?: ComponentType<TextFieldAccessoryProps>;
  error?: string;
}

export const TextField = forwardRef(function TextField(
  props: TextFieldProps,
  ref: Ref<TextInput>,
) {
  const {
    label,
    placeholder,
    helper,
    status,
    RightAccessory,
    LeftAccessory,
    HelperTextProps,
    LabelTextProps,
    error,
    style: inputStyleOverride,
    containerStyle: containerStyleOverride,
    inputWrapperStyle: inputWrapperStyleOverride,
    ...restTextInputProps
  } = props;
  const input = useRef<TextInput>(null);

  const disabled =
    restTextInputProps.editable === false || status === 'disabled';

  function focusInput() {
    if (disabled) return;
    input.current?.focus();
  }

  useImperativeHandle(ref, () => input.current);

  return (
    <TouchableOpacity
      activeOpacity={1}
      style={styles.container}
      onPress={focusInput}
      accessibilityState={{ disabled }}
    >
      {!!label && (
        <Typography
          variant="formLabel"
          text={label}
          {...LabelTextProps}
          style={styles.label}
        />
      )}

      <View style={[styles.inputWrapper, error && styles.errorInputWrapper]}>
        {!!LeftAccessory && (
          <LeftAccessory
            style={styles.accessory}
            status={status}
            editable={!disabled}
            multiline={restTextInputProps.multiline}
          />
        )}

        <TextInput
          ref={input}
          underlineColorAndroid={colors.transparent}
          textAlignVertical="top"
          placeholder={placeholder}
          placeholderTextColor={colors.textDim}
          {...restTextInputProps}
          editable={!disabled}
          style={styles.input}
        />

        {!!RightAccessory && (
          <RightAccessory
            style={styles.accessory}
            status={status}
            editable={!disabled}
            multiline={restTextInputProps.multiline}
          />
        )}
      </View>

      {(!!helper || !!error) && (
        <Typography
          variant="formHelper"
          size="xs"
          text={helper || error}
          {...HelperTextProps}
          style={[styles.helperText, error && { color: colors.angry500 }]}
        />
      )}
    </TouchableOpacity>
  );
});

const styles: {
  container: StyleProp<ViewStyle>;
  label: StyleProp<TextStyle>;
  inputWrapper: StyleProp<ViewStyle>;
  accessory: StyleProp<ViewStyle>;
  input: StyleProp<TextStyle>;
  helperText: StyleProp<TextStyle>;
  errorInputWrapper: StyleProp<ViewStyle>;
} = {
  container: {},
  label: {
    marginBottom: spacing.xs,
  },
  inputWrapper: {
    flexDirection: 'row',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 4,
    backgroundColor: colors.neutral200,
    borderColor: colors.neutral400,
    overflow: 'hidden',
  },
  errorInputWrapper: {
    backgroundColor: colors.angry100,
    borderColor: colors.angry500,
  },
  accessory: {
    marginHorizontal: spacing.xs,
    height: 40,
    justifyContent: 'center',
    alignItems: 'center',
  },
  input: {
    flex: 1,
    alignSelf: 'stretch',
    fontFamily: typography.primary.normal,
    color: colors.text,
    fontSize: 16,
    height: 24,
    paddingVertical: 0,
    paddingHorizontal: 0,
    marginVertical: spacing.xs,
    marginHorizontal: spacing.sm,
  },
  helperText: {
    marginTop: spacing.xs,
  },
};
