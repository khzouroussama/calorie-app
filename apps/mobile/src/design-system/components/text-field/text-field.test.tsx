import React from 'react';
import renderer from 'react-test-renderer';

import { TextField } from './text-field';

describe('<TextField />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<TextField value="Hellp" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
