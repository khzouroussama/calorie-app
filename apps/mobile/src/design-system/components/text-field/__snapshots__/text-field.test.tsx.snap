// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#F4F2F1",
          "borderColor": "#B6ACA6",
          "borderRadius": 4,
          "borderWidth": 1,
          "flexDirection": "row",
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <TextInput
      editable={true}
      placeholderTextColor="#564E4A"
      style={
        {
          "alignSelf": "stretch",
          "color": "#191015",
          "flex": 1,
          "fontFamily": "spaceGroteskRegular",
          "fontSize": 16,
          "height": 24,
          "marginHorizontal": 12,
          "marginVertical": 8,
          "paddingHorizontal": 0,
          "paddingVertical": 0,
        }
      }
      textAlignVertical="top"
      underlineColorAndroid="rgba(0, 0, 0, 0)"
      value="Hellp"
    />
  </View>
</View>
`;
